Match metavariable in dot access for php

emjin opened the pull request:
Distinguish metavar from var when parsing
    
Because php has $X as a legal dynamic variable, but we are choosing to interpret that as a metavariable in our pattern, it is necessary to distinguish whether a variable of the form $X is a variable or metavariable at the parser level. Without this distinction, it is later assumed to be a dynamic variable.
    
Added a MetaVar to the cst and Metavar (which later is unified with Id) to the ast. When a variable name is all caps (currently tested by checking it against uppercase_ascii), it is assumed to be a metavariable.
    
Test plan:
    Run with make test. See corresponding semgrep branch for the new tests. Can also run one of the new sgrep files with the -dump_pattern flag, expect EId("$X", ...), not EDynamic(Id("$X", ...), ...).

(Sorry about the terrible commit message "resolved conflict". I didn't understand something with github
The above was my original message)

emjin committed:
Resolved conflict

emjin committed:
Fixed whitespace

aryx commented on lang_php/analyze/ast_php.ml:
```diff
@@ -169,6 +169,7 @@ type expr =
    * The string contains the '$'.
   *)
   | Var of var
+  | Metavar of var (* sgrep-ext *)
```
You need to add the comment about why we need that in PHP whereas we just hijack Id in the other languages.

aryx replied:
An alternative is to keep Var, like before, but use at more places the check AST_generic_helpers.is_metavar_name
to distinguish the two, especially if we are in Flag_parsing.sgrep_mode

aryx commented on lang_php/analyze/php_to_generic.ml:
```diff
@@ -263,11 +265,16 @@ and expr =
   | Obj_get (v1, t, Id [v2]) ->
       let v1 = expr v1 and v2 = ident v2 in
       G.DotAccess (v1, t, G.EId (v2, G.empty_id_info()))
+  | Obj_get (v1, t, Metavar v2) ->
```
ok, got it. This is the main issue. But again an alternative is to add a case like
| Obj_get (v1, t, Var ((s, _t)) when AST_generic_helpers.is_metavar_name s && !Flag_parsing.sgrep_mode -> 
 ...

It's just that in sgrep_mode, we really want to treat those Var like Id, which we do when converting to the generic AST, but the problem is that each time we do deeper match where we use Id, then we need also to catch Var (s,_) when is_metavar_name && sgrep_mode.

aryx commented on lang_php/analyze/php_to_generic.ml:
```diff
@@ -263,11 +265,16 @@ and expr =
       G.DotAccess (v1, t, G.EDynamic v2)
   | Class_get (v1, t, Id [v2]) -> let v1 = expr v1 and v2 = ident v2 in
       G.DotAccess (v1, t, G.EId (v2, G.empty_id_info()))
+  | Class_get (v1, t, Metavar v2) -> let v1 = expr v1 and v2 = ident v2 in
```
yep, same here, there is a complex pattern where Id is matched deep, and we want to also handle
the Var when is_metavar_name && sgrep_mode.

Maybe there are more places in php_to_generic.ml where we match Id deeply; Check for it.

aryx replied:
I've seen one for the Global case for example (and UseOuterDecl)
maybe we also need to handle specially Var (or your Metavar) there.

aryx replied:
There's also one in the code to transform attribute. Basically, each time we match Id, we need
to add a case for Var when the var is a metavar.

aryx commented on lang_php/parsing/parser_php.mly:
```diff
@@ -74,7 +74,10 @@ let mk_param s =
-  | _ -> IdVar (DName(s, tok))
+  | _ ->
+    (* assume variable is a metavariable if it is all uppercase *)
+    if String.uppercase_ascii s = s then MetaVar (DName(s, tok))
```
You can use AST_generic_helpers.is_metavar_name ("$" ^ s)

aryx replied:
And you need to make sure we do that only when Flag_parsing.sgrep_mode is true.

As I said, I'd rather not add a Metavar but at more places where we match Id, we should add match
for Var with the is_metavar_name && sgrep_mode attached to it.

emjin committed:
Changed approach to fixing this bug

In php_to_generic, convert Id [v1] to G.Id, since there are no qualifications. This means that Id [v1] can also be used to represent metavariables. The CST still uses metavar as a construct, since Ids are separated into several possible constructs, but in ast_php_build MetaVar is unified with Id.

Test plan: make test

emjin committed:
Use is_metavar_name and sgrep_mode

Used existing method AST_generic_helpers.is_metavar_name to test for uppercase, and also check for sgrep_mode

Note that AST_generic_helpers directions you to Metavars_generic, but there is no mli for that

emjin committed:
Added comment about MetaVar

aryx commented:
The CI failure are due to wrong indentation. You need to setup correctly pre-commit
to run when you commit, so that it runs ocp-indent for you.
You can also invoke it manually with pre-commit run -a, but if you set pre-commit correctly it should
run automatiically when you commit and tell you, without having to wait for CI, if there is a problem.

aryx commented on lang_php/analyze/ast_php_build.ml:
```diff
@@ -315,6 +315,7 @@ and expr env = function
   | Id n -> A.Id (name env n)
 
   | IdVar dn -> A.Var (dname dn)
+  | MetaVar dn -> A.Id [dname dn]
```
will the resulting Id will have a $ in its string when it comes from a Metavar?

aryx reviewed (changes requested):
Fix ocp-indent and resubmit.
Looks good otherwise.

aryx commented:
Closing that one too, in preference to #385
